--- a/include/acus/assembler/assembler.h
+++ b/include/acus/assembler/assembler.h
@@ -614,37 +614,52 @@
   template <ws::Bool16Operand W>    Bool16Result<W>  bool16Destructive(W const &val);
 
   // Logical functions (assembler_logical.{cc,tpp}
+  void boolDestructive(Cell target, Cell tmp);
+  void notDestructive(Cell x, Cell tmp);
+  void orDestructive(Cell x, Cell y);
+  void orDestructive(Cell x, Cell y, Cell tmp);
+  void andDestructive(Cell x, Cell y, Cell tmp);
+  void xorDestructive(Cell x, Cell y, Cell tmp);
+  void nandDestructive(Cell x, Cell y, Cell tmp);
+  void norDestructive(Cell x, Cell y, Cell tmp);
+  void xnorDestructive(Cell x, Cell y, Cell tmp);
+
   template <ws::SingleAndScratch W> Data8Result<W>  notDestructive(W const &op);
   template <ws::DoubleCell W>       Bool16Result<W> not16Destructive(W const &op);
-  template <ws::OrOperand W>         Data8Result<W>  orDestructive(W const &lhs, SingleCell const &rhs);
-  template <ws::LogicInternalOperand W> Data8Result<W>  orInternalDestructive(W const &lhs);
-  template <ws::SingleAndScratch W>  Data8Result<W>  andDestructive(W const &lhs, SingleCell const &rhs);
-  template <ws::SingleAndScratch W>  Data8Result<W>  xorDestructive(W const &lhs, SingleCell const &rhs);
-  template <ws::SingleAndScratch W>  Data8Result<W>  nandDestructive(W const &lhs, SingleCell const &rhs);
-  template <ws::SingleCell W>        Data8Result<W>  norDestructive(W const &lhs, SingleCell const &rhs);
-  template <ws::SingleAndScratch W>  Data8Result<W>  xnorDestructive(W const &lhs, SingleCell const &rhs);
-
-
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> logic16Destructive(L const &lhs, R const &rhs, auto &&logicOp);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> or16Destructive(L const &lhs, R const &rhs);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> and16Destructive(L const &lhs, R const &rhs);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> nor16Destructive(L const &lhs, R const &rhs);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> nand16Destructive(L const &lhs, R const &rhs);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> xor16Destructive(L const &lhs, R const &rhs);
-  template <ws::Logic16Operand L, ws::Logic16Operand R> Bool16Result<L> xnor16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> orDestructive(W const &lhs, SingleCell const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> andDestructive(W const &lhs, SingleCell const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> xorDestructive(W const &lhs, SingleCell const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> nandDestructive(W const &lhs, SingleCell const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> norDestructive(W const &lhs, SingleCell const &rhs);
+  template <ws::BinaryLogic8Operand W> Data8Result<W> xnorDestructive(W const &lhs, SingleCell const &rhs);
+
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R>
+  Bool16Result<L> logic16Destructive(L const &lhs, R const &rhs, auto &&logicOp);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> or16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> and16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> nor16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> nand16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> xor16Destructive(L const &lhs, R const &rhs);
+  template <ws::BinaryLogic16Operand L, ws::BinaryLogic16Operand R> Bool16Result<L> xnor16Destructive(L const &lhs, R const &rhs);
 
   // Comparisons (assembler_comparisons.{cc,tpp}
+  void eqDestructive(Cell x, Cell y);
+  void lessDestructive(Cell x, Cell y, Cell tmp);
+  void lessOrEqualDestructive(Cell x, Cell y, Cell tmp);
+  void greaterDestructive(Cell x, Cell y, Cell tmp);
+  void greaterOrEqualDestructive(Cell x, Cell y, Cell tmp);
+
   template <ws::Compare8Operand W> Data8Result<W> eqDestructive(W const &lhs, SingleCell const &rhs);
   template <ws::Compare8Operand W> Data8Result<W> lessDestructive(W const &lhs, SingleCell const &rhs);
   template <ws::Compare8Operand W> Data8Result<W> lessOrEqualDestructive(W const &lhs, SingleCell const &rhs);
   template <ws::Compare8Operand W> Data8Result<W> greaterDestructive(W const &lhs, SingleCell const &rhs);
   template <ws::Compare8Operand W> Data8Result<W> greaterOrEqualDestructive(W const &lhs, SingleCell const &rhs);
 
-  template <ws::DoubleAndScratch L>                     Bool16Result<L> eq16Destructive(L const &lhs, DoubleCell const &rhs);
-  template <ws::Compare16Lhs L, ws::DoubleAndScratch R> Bool16Result<L> less16Destructive(L const &lhs, R const &rhs);
-  template <ws::Compare16Lhs L, ws::DoubleAndScratch R> Bool16Result<L> lessOrEqual16Destructive(L const &lhs, R const &rhs);
-  template <ws::Compare16Lhs L, ws::DoubleAndScratch R> Bool16Result<L> greater16Destructive(L const &lhs, R const &rhs);
-  template <ws::Compare16Lhs L, ws::DoubleAndScratch R> Bool16Result<L> greaterOrEqual16Destructive(L const &lhs, R const &rhs);
+  template <ws::Eq16Operand L>      Bool16Result<L> eq16Destructive(L const &lhs, DoubleCell const &rhs);
+  template <ws::Compare16Operand L> Bool16Result<L> less16Destructive(L const &lhs, DoubleCell const &rhs);
+  template <ws::Compare16Operand L> Bool16Result<L> lessOrEqual16Destructive(L const &lhs, DoubleCell const &rhs);
+  template <ws::Compare16Operand L> Bool16Result<L> greater16Destructive(L const &lhs, DoubleCell const &rhs);
+  template <ws::Compare16Operand L> Bool16Result<L> greaterOrEqual16Destructive(L const &lhs, DoubleCell const &rhs);
 
   // Frame Navigation (assembler_framenav.cc)
   void resetOrigin();
--- a/include/acus/core/data.h
+++ b/include/acus/core/data.h
@@ -763,13 +763,35 @@
       >;
 
     template <typename W>
-    concept Compare8Operand = impl::isWorkspace<W> && W::template Satisfies<
+    concept BinaryLogic8Operand = impl::IsWorkspace<W> && W::template Satisfies<
       Data<>,
       Clobber,
       Scratch
     >;
     
     template <typename W>
+    concept Bool16Operand = DoubleCell<W>;
+
+    template <typename W>
+    concept BinaryLogic16Operand = DoubleCell<W> && W::template Satisfies<
+      DataCells<2>,
+      Scratch
+    >;
+
+    template <typename W>
+    concept Compare8Operand = impl::IsWorkspace<W> && W::template Satisfies<
+      Data<>,
+      Clobber,
+      Scratch
+    >;
+
+    template <typename W>
+    concept Eq16Operand = DoubleCell<W> && W::template Satisfies<
+      DataCells<2>,
+      ScratchCells<2>
+    >;
+
+    template <typename W>
     concept Compare16Operand = DoubleCell<W> && W::template Satisfies<
       DataCells<2>,
       ScratchCells<5>
--- a/src/assembler/assembler_algorithms.cc
+++ b/src/assembler/assembler_algorithms.cc
@@ -105,7 +105,7 @@
 }
 
 void Assembler::moveField(Cell from, Cell to) {
-  if (from == to) return;
+  if (from.offset == to.offset && from.field == to.field) return;
 
   zeroCell(to);
   loop(from, [&]{
@@ -118,12 +118,12 @@
   for (Cell const &c: to) zeroCell(c);
   loop(from, [&]{
     dec(from);
-    for (Cell const &c: to) inc(to); // TODO: optimize the order
+    for (Cell const &c: to) inc(c); // TODO: optimize the order
   });
 }
 
 void Assembler::copyField(Cell from, Cell to, Cell tmp) {
-  if (from == to) return;
+  if (from.offset == to.offset && from.field == to.field) return;
   moveField(from, {to, tmp});
   moveField(tmp, from);
 }
--- a/src/assembler/assembler_comparisons.tpp
+++ b/src/assembler/assembler_comparisons.tpp
@@ -2,9 +2,9 @@
   template <ws::Compare8Operand W>                                      \
   Assembler::Data8Result<W> Assembler::op##Destructive(W const &lhs, SingleCell const &rhs) { \
     auto const [x, y, tmp] = lhs.template cells<3>();                   \
-    moveField(rhs, y);                                                  \
+    moveField(rhs[0], y);                                               \
     op##Destructive(x, y, tmp);                                         \
-    return lhs.template transformed<ws::Replace<1, Data<0>>>();         \
+    return lhs.template transformed<ws::Replace<1, ws::Data<0>>>();     \
   }
 
 
@@ -15,12 +15,12 @@
 
 #undef COMPARE_8_IMPL
 
-template <ws::DoubleCell W>                                      
+template <ws::Compare8Operand W>
 Assembler::Data8Result<W> Assembler::eqDestructive(W const &lhs, SingleCell const &rhs) { 
   auto const [x, y] = lhs.template cells<2>();
-  moveField(rhs, y);                                                  
+  moveField(rhs[0], y);
   eqDestructive(x, y);                                         
-  return lhs.template transformed<ws::Replace<1, Data<0>>>();         
+  return lhs.template transformed<ws::Replace<1, ws::Data<0>>>();
 }
 
 template <ws::Eq16Operand L>
@@ -36,14 +36,14 @@
 
   eqDestructive(w["xLow"],  w["yLow"]);
   eqDestructive(w["xHigh"], w["yHigh"]);
-  andDestructive(w["xLow"], w["xHigh"]);
+  andDestructive(w["xLow"], w["xHigh"], w["yHigh"]);
   return lhs.template transformed<ws::Replace<0, ws::Data<>>,
-                                  ws::Replace<0, ws::Data<>>>();
+                                  ws::Replace<1, ws::Data<0>>>();
 }
   
 
-template <ws::Compare16Operand L, ws::DoubleCell R>
-Assembler::Bool16Result<L> Assembler::less16Destructive(L const &lhs, R const &rhs) {
+template <ws::Compare16Operand L>
+Assembler::Bool16Result<L> Assembler::less16Destructive(L const &lhs, DoubleCell const &rhs) {
   auto [xLow, xHigh] = lhs.template cells<2>();
   auto [yLow, yHigh] = rhs.template cells<2>();
 
@@ -60,9 +60,8 @@
   lessDestructive(w["xLow"], w["yLow"], w["tmp"]);
   eqDestructive(w["xHigh2"], w["yHigh2"]);
 
-  w.renameAll("highLess", "tmp1", "tmp2", "lowLess", "tmp3", "highEqual", "tmp3");
-  andDestructive(w["lowLess"], w["highEqual"], "tmp2");
-  orDestructive(w["highLess"], w["lowLess"]);
+  andDestructive(w["xLow"], w["xHigh2"], w["tmp"]);
+  orDestructive(w["xHigh1"], w["xLow"]);
 
   return lhs.template transformed<
     ws::Replace<0, ws::Data<>>,
@@ -70,8 +69,8 @@
   >();
 }
 
-template <ws::Compare16Lhs L, ws::DoubleAndScratch R>
-Assembler::Bool16Result<L> Assembler::greater16Destructive(L const &lhs, R const &rhs) {
+template <ws::Compare16Operand L>
+Assembler::Bool16Result<L> Assembler::greater16Destructive(L const &lhs, DoubleCell const &rhs) {
   auto [xLow, xHigh] = lhs.template cells<2>();
   auto [yLow, yHigh] = rhs.template cells<2>();
 
@@ -88,9 +87,8 @@
   greaterDestructive(w["xLow"], w["yLow"], w["tmp"]);
   eqDestructive(w["xHigh2"], w["yHigh2"]);
 
-  w.renameAll("highGreater", "tmp1", "tmp2", "lowGreater", "tmp3", "highEqual", "tmp3");
-  andDestructive(w["lowGreater"], w["highEqual"], "tmp2");
-  orDestructive(w["highGreater"], w["lowGreater"]);
+  andDestructive(w["xLow"], w["xHigh2"], w["tmp"]);
+  orDestructive(w["xHigh1"], w["xLow"]);
 
   return lhs.template transformed<
     ws::Replace<0, ws::Data<>>,
@@ -98,16 +96,16 @@
   >();
 }
 
-template <ws::Compare16Lhs L, ws::DoubleAndScratch R>
-Assembler::Bool16Result<L> Assembler::lessOrEqual16Destructive(L const &lhs, R const &rhs) {
+template <ws::Compare16Operand L>
+Assembler::Bool16Result<L> Assembler::lessOrEqual16Destructive(L const &lhs, DoubleCell const &rhs) {
   auto result = greater16Destructive(lhs, rhs);
-  notDestructive(result);
+  notDestructive(result[0], result[1]);
   return result;
 }
 
-template <ws::Compare16Lhs L, ws::DoubleAndScratch R>
-Assembler::Bool16Result<L> Assembler::greaterOrEqual16Destructive(L const &lhs, R const &rhs) {
+template <ws::Compare16Operand L>
+Assembler::Bool16Result<L> Assembler::greaterOrEqual16Destructive(L const &lhs, DoubleCell const &rhs) {
   auto result = less16Destructive(lhs, rhs);
-  notDestructive(result);
+  notDestructive(result[0], result[1]);
   return result;
 }
--- a/src/assembler/assembler_logical.cc
+++ b/src/assembler/assembler_logical.cc
@@ -160,10 +160,10 @@
 
 void Assembler::boolDestructive(Cell target, Cell tmp) {
   loop(target, [&]{
-    zeroCell(target)
+    zeroCell(target);
     inc(tmp);
   });
-  addDestructive(target, tmp);
+  moveField(tmp, target);
 }
   
 void Assembler::notDestructive(Cell x, Cell tmp) {
@@ -173,7 +173,7 @@
     zeroCell(x);
   });
   // x = 0, tmp = not(x)
-  addDestructive(x, tmp);
+  moveField(tmp, x);
 }
 
 void Assembler::orDestructive(Cell lhs, Cell rhs) {
@@ -188,6 +188,10 @@
   });
 }
 
+void Assembler::orDestructive(Cell lhs, Cell rhs, Cell) {
+  orDestructive(lhs, rhs);
+}
+
 void Assembler::andDestructive(Cell x, Cell y, Cell tmp) {
   addConst(tmp, 2);  // tmp is scratch -> known 0
   loop(x, [&]{
--- a/src/assembler/assembler_logical.tpp
+++ b/src/assembler/assembler_logical.tpp
@@ -1,5 +1,3 @@
-
-
 template <ws::SingleAndScratch W>
 Assembler::Data8Result<W> Assembler::boolDestructive(W const &op) {
   constexpr size_t scratch = W::template ScratchOffset<1>;
@@ -9,7 +7,7 @@
 
 template <ws::Bool16Operand W>
 Assembler::Bool16Result<W> Assembler::bool16Destructive(W const &op) {
-  orInternalDestructive(op);
+  orDestructive(op[0], op[1]);
   return op.template transformed<
     ws::Replace<0, ws::Data<>>,
     ws::Replace<1, ws::Data<0>>
@@ -35,14 +33,14 @@
 Assembler::Bool16Result<L> Assembler::logic16Destructive(L const &lhs, R const &rhs, auto &&logicOp) {
   auto lhsBool = bool16Destructive(lhs);
   auto rhsBool = bool16Destructive(rhs);
-  return logicOp(lhsBool, lhsBool);
+  return logicOp(lhsBool, rhsBool);
 }
 
 #define LOGIC_8_IMPL(op)                                                \
   template <ws::BinaryLogic8Operand W>                                  \
   Assembler::Data8Result<W> Assembler::op##Destructive(W const &lhs, SingleCell const &rhs) { \
     auto const [x, y, tmp] = lhs.template cells<3>();                   \
-    moveField(rhs, y);                                                  \
+    moveField(rhs[0], y);                                               \
     op##Destructive(x, y, tmp);                                         \
     return lhs.template transformed<ws::Replace<1, ws::Data<0>>>();     \
   }
@@ -72,4 +70,3 @@
 
 #undef LOGIC_8_IMPL
 #undef LOGIC_16_IMPL
-
